From a9420c9a607c653222a48d5ba993e86bae1c6ff3 Mon Sep 17 00:00:00 2001
From: dianlujitao <dianlujitao@lineageos.org>
Date: Thu, 14 Feb 2019 13:28:42 +0800
Subject: [PATCH 2/9] sepolicy: Dynamically build trust policy into
 system/vendor

 * Introduce a new board flag TARGET_USES_PREBUILT_VENDOR_SEPOLICY and
   two sepolicy variants: dynamic and priv_vendor
 * When TARGET_USES_PREBUILT_VENDOR_SEPOLICY=true, dynamic and
   priv_vendor act as private policy, and vendor policy is excluded in
   order to avoid conflicts (it's not packaged into final builds
   anyway), When the flag is not set, dynamic acts as vendor policy and
   priv_vendor is execluded.
 * Trust policy in priv_vendor is taken from xiaomi sdm845-common.

Change-Id: I8bfd078d6064616c88e2c58a9fa3aa045dddf303
---
 common/dynamic/file.te                          | 1 +
 common/dynamic/genfs_contexts                   | 1 +
 common/{public => dynamic}/hal_lineage_trust.te | 0
 common/dynamic/hwservice.te                     | 1 +
 common/dynamic/hwservice_contexts               | 1 +
 common/priv_vendor/file_contexts                | 2 ++
 common/priv_vendor/hal_lineage_trust_default.te | 8 ++++++++
 common/private/genfs_contexts                   | 2 --
 common/private/hwservice_contexts               | 1 -
 common/public/file.te                           | 2 --
 common/public/hwservice.te                      | 1 -
 common/sepolicy.mk                              | 7 +++++++
 qcom/sepolicy.mk                                | 2 ++
 13 files changed, 23 insertions(+), 6 deletions(-)
 create mode 100644 common/dynamic/file.te
 create mode 100644 common/dynamic/genfs_contexts
 rename common/{public => dynamic}/hal_lineage_trust.te (100%)
 create mode 100644 common/dynamic/hwservice.te
 create mode 100644 common/dynamic/hwservice_contexts
 create mode 100644 common/priv_vendor/file_contexts
 create mode 100644 common/priv_vendor/hal_lineage_trust_default.te

diff --git a/common/dynamic/file.te b/common/dynamic/file.te
new file mode 100644
index 0000000..c77d9ec
--- /dev/null
+++ b/common/dynamic/file.te
@@ -0,0 +1 @@
+type proc_deny_new_usb, fs_type, proc_type;
diff --git a/common/dynamic/genfs_contexts b/common/dynamic/genfs_contexts
new file mode 100644
index 0000000..60cf2c6
--- /dev/null
+++ b/common/dynamic/genfs_contexts
@@ -0,0 +1 @@
+genfscon proc /sys/kernel/deny_new_usb u:object_r:proc_deny_new_usb:s0
diff --git a/common/public/hal_lineage_trust.te b/common/dynamic/hal_lineage_trust.te
similarity index 100%
rename from common/public/hal_lineage_trust.te
rename to common/dynamic/hal_lineage_trust.te
diff --git a/common/dynamic/hwservice.te b/common/dynamic/hwservice.te
new file mode 100644
index 0000000..7ca4141
--- /dev/null
+++ b/common/dynamic/hwservice.te
@@ -0,0 +1 @@
+type hal_lineage_trust_hwservice, hwservice_manager_type;
diff --git a/common/dynamic/hwservice_contexts b/common/dynamic/hwservice_contexts
new file mode 100644
index 0000000..6cb1181
--- /dev/null
+++ b/common/dynamic/hwservice_contexts
@@ -0,0 +1 @@
+vendor.lineage.trust::IUsbRestrict                   u:object_r:hal_lineage_trust_hwservice:s0
diff --git a/common/priv_vendor/file_contexts b/common/priv_vendor/file_contexts
new file mode 100644
index 0000000..9fdebb5
--- /dev/null
+++ b/common/priv_vendor/file_contexts
@@ -0,0 +1,2 @@
+# Trust HAL
+/system/bin/hw/lineage\.trust@1\.0-service u:object_r:hal_lineage_trust_default_exec:s0
diff --git a/common/priv_vendor/hal_lineage_trust_default.te b/common/priv_vendor/hal_lineage_trust_default.te
new file mode 100644
index 0000000..0f393c9
--- /dev/null
+++ b/common/priv_vendor/hal_lineage_trust_default.te
@@ -0,0 +1,8 @@
+type hal_lineage_trust_default, coredomain, domain;
+hal_server_domain(hal_lineage_trust_default, hal_lineage_trust)
+
+type hal_lineage_trust_default_exec, exec_type, file_type;
+init_daemon_domain(hal_lineage_trust_default)
+
+# Talk to the binder device node
+allow hal_lineage_trust_default binder_device:chr_file rw_file_perms;
diff --git a/common/private/genfs_contexts b/common/private/genfs_contexts
index 62ba3e3..09d7df6 100644
--- a/common/private/genfs_contexts
+++ b/common/private/genfs_contexts
@@ -11,5 +11,3 @@ genfscon sysfs /devices/virtual/graphics/fb0/sre u:object_r:sysfs_livedisplay_tu
 genfscon sysfs /devices/virtual/graphics/fb0/reading_mode u:object_r:sysfs_livedisplay_tuneable:s0
 
 genfscon sysfs /devices/virtual/timed_output/vibrator u:object_r:sysfs_vibrator:s0
-
-genfscon proc /sys/kernel/deny_new_usb u:object_r:proc_deny_new_usb:s0
diff --git a/common/private/hwservice_contexts b/common/private/hwservice_contexts
index 58ffb72..71d03ed 100644
--- a/common/private/hwservice_contexts
+++ b/common/private/hwservice_contexts
@@ -12,4 +12,3 @@ vendor.lineage.touch::IGloveMode                     u:object_r:hal_lineage_touc
 vendor.lineage.touch::IKeyDisabler                   u:object_r:hal_lineage_touch_hwservice:s0
 vendor.lineage.touch::IStylusMode                    u:object_r:hal_lineage_touch_hwservice:s0
 vendor.lineage.touch::ITouchscreenGesture            u:object_r:hal_lineage_touch_hwservice:s0
-vendor.lineage.trust::IUsbRestrict                   u:object_r:hal_lineage_trust_hwservice:s0
diff --git a/common/public/file.te b/common/public/file.te
index 9b6c35e..45564dc 100644
--- a/common/public/file.te
+++ b/common/public/file.te
@@ -1,3 +1 @@
-type proc_deny_new_usb, fs_type, proc_type;
-
 type sysfs_livedisplay_tuneable, fs_type, sysfs_type;
diff --git a/common/public/hwservice.te b/common/public/hwservice.te
index 62a4f87..3676f11 100644
--- a/common/public/hwservice.te
+++ b/common/public/hwservice.te
@@ -1,3 +1,2 @@
 type hal_lineage_livedisplay_hwservice, hwservice_manager_type;
 type hal_lineage_touch_hwservice, hwservice_manager_type;
-type hal_lineage_trust_hwservice, hwservice_manager_type;
diff --git a/common/sepolicy.mk b/common/sepolicy.mk
index 7e8299a..8c39b97 100644
--- a/common/sepolicy.mk
+++ b/common/sepolicy.mk
@@ -9,5 +9,12 @@ BOARD_PLAT_PUBLIC_SEPOLICY_DIR += \
 BOARD_PLAT_PRIVATE_SEPOLICY_DIR += \
     device/lineage/sepolicy/common/private
 
+ifeq ($(TARGET_USES_PREBUILT_VENDOR_SEPOLICY), true)
+BOARD_PLAT_PRIVATE_SEPOLICY_DIR += \
+    device/lineage/sepolicy/common/dynamic \
+    device/lineage/sepolicy/common/priv_vendor
+else
 BOARD_SEPOLICY_DIRS += \
+    device/lineage/sepolicy/common/dynamic \
     device/lineage/sepolicy/common/vendor
+endif
diff --git a/qcom/sepolicy.mk b/qcom/sepolicy.mk
index c65816c..d6fbd9a 100644
--- a/qcom/sepolicy.mk
+++ b/qcom/sepolicy.mk
@@ -6,6 +6,8 @@
 BOARD_PLAT_PRIVATE_SEPOLICY_DIR += \
     device/lineage/sepolicy/qcom/private
 
+ifneq ($(TARGET_USES_PREBUILT_VENDOR_SEPOLICY), true)
 BOARD_SEPOLICY_DIRS += \
     device/lineage/sepolicy/qcom/common \
     device/lineage/sepolicy/qcom/$(TARGET_BOARD_PLATFORM)
+endif
-- 
2.17.1

